fairworkflows/fairworkflow.py

```diff
@@ -86,24 +86,29 @@ def from_function(cls, func: Callable):
                              'use is_fairworkflow decorator to mark it.')
 
     @classmethod
-    def from_noodles_promise(cls, noodles_promise, description: str = None, label: str = None,
-                 is_pplan_plan: bool = True, derived_from=None):
+    def from_noodles_promise(cls, workflow_level_promise, step_level_promise, description: str = None, label: str =
+                             None, is_pplan_plan: bool = True, derived_from=None):
+        """
+
+        Args:
+            workflow_level_promise: Noodles workflow_level_promise at the workflow function level.
+                This excludes the individual steps, but we need it to do proper execution.
+            step_level_promise: Promise at the steps level. This includes the individual steps as
+                nodes but does not bind them together. We use this to extract step information from
+                it.
+        """
         self = cls(description=description, label=label, is_pplan_plan=is_pplan_plan,
                    derived_from=derived_from, language=LINGSYS_PYTHON)
-        self.noodles_promise = noodles_promise
+        self.workflow_level_promise = workflow_level_promise
 
-        workflow = noodles.get_workflow(self.noodles_promise)
+        workflow = noodles.get_workflow(step_level_promise)
 
-        # Exclude the root, because that is the workflow definition itself and not a step
-        steps_dict = {i: n.foo._fairstep for i, n in workflow.nodes.items() if i != workflow.root}
+        steps_dict = {i: n.foo._fairstep for i, n in workflow.nodes.items()}
```

Comment:
I confess I've sort of forgotten what the reasoning was here - so we're now including the workflow 'root' step?

Reply:
We're now using the `step_level_promise` which does not have the reference to the function defining the workflow itself. The root node is thus one of the steps and we want to include it.